pkg/volumes/do/volumes.go

```diff
@@ -0,0 +1,345 @@
+			}
+
+			if len(doVolume.DropletIDs) == 1 {
+				vol.AttachedTo = strconv.Itoa(doVolume.DropletIDs[0])
+				vol.LocalDevice = getLocalDeviceName(&doVolume)
+			}
+
+			glog.V(2).Infof("Matching DO Volume found with name=%s ID=%s etcdname=%s mountname=%s", doVolume.Name, doVolume.ID, vol.EtcdName, vol.MountName)
+
+			myvolumes = append(myvolumes, vol)
+		}
+
+	}
+
+	return myvolumes, nil
+}
+
+// FindMountedVolume implements Volumes::FindMountedVolume
+func (a *DOVolumes) FindMountedVolume(volume *volumes.Volume) (string, error) {
+	device := volume.LocalDevice
+
+	_, err := os.Stat(volumes.PathFor(device))
+	if err == nil {
+		return device, nil
+	}
+
+	if !os.IsNotExist(err) {
+		return "", fmt.Errorf("error checking for device %q: %v", device, err)
+	}
+
+	return "", fmt.Errorf("error checking for device %q: %v", device, err)
+}
+
+// AttachVolume attaches the specified volume to this instance, returning the mountpoint & nil if successful
+func (a *DOVolumes) AttachVolume(volume *volumes.Volume) error {
+	for {
+		action, _, err := a.DigiCloud.Client.StorageActions.Attach(context.TODO(), volume.ProviderID, a.dropletID)
+		if err != nil {
+			return fmt.Errorf("error attaching volume: %s", err)
+		}
+
+		if action.Status != godo.ActionInProgress && action.Status != godo.ActionCompleted {
+			return fmt.Errorf("invalid status for digitalocean volume: %s", volume.ProviderID)
+		}
+
+		doVolume, err := getVolumeByID(a.DigiCloud, volume.ProviderID)
+		if err != nil {
+			return fmt.Errorf("error getting volume status: %s", err)
+		}
+
+		if len(doVolume.DropletIDs) == 1 {
+			if doVolume.DropletIDs[0] != a.dropletID {
+				return fmt.Errorf("digitalocean volume %s is attached to another droplet", doVolume.ID)
+			}
+
+			volume.LocalDevice = getLocalDeviceName(doVolume)
+			return nil
+		}
+
+		time.Sleep(10 * time.Second)
+	}
+}
+
+func (a *DOVolumes) MyIP() (string, error) {
+	addr, err := getMetadata(dropletInternalIPMetadataURL)
+	if err != nil {
+		return "", err
+	}
+
+	return addr, nil
+}
+
+func getMetadataRegion() (string, error) {
+	return getMetadata(dropletRegionMetadataURL)
+}
+
+func getMetadataDropletName() (string, error) {
+	return getMetadata(dropletNameMetadataURL)
+}
+
+func getMetadataDropletID() (string, error) {
+	return getMetadata(dropletIDMetadataURL)
+}
+
+func getMetadata(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("droplet metadata returned non-200 status code: %d", resp.StatusCode)
+	}
+
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bodyBytes), nil
+}
+
+// NewCloud returns a Cloud, expecting the env var DIGITALOCEAN_ACCESS_TOKEN
+// NewCloud will return an err if DIGITALOCEAN_ACCESS_TOKEN is not defined
+func NewCloud(region string) (*DigiCloud, error) {
+	accessToken := os.Getenv("DIGITALOCEAN_ACCESS_TOKEN")
```

Comment:
how you are going to pass these when using Kops?

Reply:
I think etcd-manager pod should have these env variables, but it does not have anything https://github.com/kubernetes/kops/blob/master/pkg/model/components/etcdmanager/model.go#L184

Reply:
I haven't raised a PR for the KOPS change, but here is what I am doing.
https://github.com/kubernetes/kops/compare/master...srikiz:DO-7148-EtcdManagerUpdates?expand=1#diff-cc02d582f0c8a1fc88c16dedef2e30edR443

Reply:
The manifest is overridden by passing the additional environment variables.

Reply:
could you remove that if digitalocean part, instead check does env variables like `S3_ENDPOINT` exist. If it exist append to pod?

I am asking this because I am using `S3_ENDPOINT`, `S3_ACCESS_KEY_ID` and `S3_SECRET_ACCESS_KEY` variables in OpenStack and AWS kops clusters. So it is not cloudprovider related.

Reply:
👍

Reply:
Check https://github.com/kubernetes/kops/pull/7395 I added it there